File: frontend/src/oauthIntegrationStatus.ts
import { TEMPLATE_MIRO } from "./admin/constants";
import type { ProviderAppOut, ProviderInstanceOut } from "./types";

export type OAuthInstanceLike = {
  authorization_endpoint?: string | null;
  token_endpoint?: string | null;
  settings?: Record<string, unknown>;
};

export function oauthInstanceFieldsFromApp(app: ProviderAppOut): OAuthInstanceLike {
  return {
    authorization_endpoint: app.oauth_authorization_endpoint,
    token_endpoint: app.oauth_token_endpoint,
    settings: app.oauth_instance_settings ?? {},
  };
}

export function integrationNeedsTenantForStatus(templateKey: string | null): boolean {
  if (templateKey === TEMPLATE_MIRO || templateKey === null) return false;
  return true;
}

export function oauthIntegrationConfigured(
  app: ProviderAppOut,
  instance: OAuthInstanceLike | ProviderInstanceOut | undefined,
  options?: { needsTenant?: boolean },
): { ok: boolean; reason: string | null } {
  const needsTenant = options?.needsTenant ?? integrationNeedsTenantForStatus(app.template_key);
  const inst = instance ?? oauthInstanceFieldsFromApp(app);
  const settings = (inst.settings ?? app.oauth_instance_settings ?? {}) as Record<string, unknown>;
  const tenantOk =
    !needsTenant || Boolean((settings.tenant_id as string | undefined)?.toString().trim());
  const authz = String(inst.authorization_endpoint ?? app.oauth_authorization_endpoint ?? "").trim();
  const tok = String(inst.token_endpoint ?? app.oauth_token_endpoint ?? "").trim();
  const pkce = Boolean((settings as { use_pkce?: boolean }).use_pkce);
  if (!tenantOk) {
    return { ok: false, reason: "Directory (tenant) not set" };
  }
  if (!String(app.client_id ?? "").trim()) {
    return { ok: false, reason: "Client ID missing" };
  }
  if (!authz) {
    return { ok: false, reason: "Authorize URL missing" };
  }
  if (!tok) {
    return { ok: false, reason: "Token URL missing" };
  }
  if (!app.has_client_secret && !pkce) {
    return { ok: false, reason: "Client secret or PKCE required" };
  }
  return { ok: true, reason: null };
}

export function integrationCardStatus(
  app: ProviderAppOut,
  instance: OAuthInstanceLike | ProviderInstanceOut | undefined,
  needsTenant: boolean,
): { label: string; tone: "neutral" | "success" | "warn" | "danger" } {
  if (!app) return { label: "Not configured", tone: "neutral" };
  const cfg = oauthIntegrationConfigured(app, instance, { needsTenant });
  if (!cfg.ok) return { label: "Not configured", tone: "neutral" };
  const instanceEnabled =
    instance && typeof (instance as ProviderInstanceOut).is_enabled === "boolean"
      ? (instance as ProviderInstanceOut).is_enabled
      : true;
  if (!app.is_enabled || !instanceEnabled) {
    return { label: "Disabled", tone: "danger" };
  }
  return { label: "Active", tone: "success" };
}

File: backend/app/oauth_integration_status.py
from __future__ import annotations

from app.models import ProviderApp, ProviderInstance
from app.security import loads_json


def _use_pkce(instance: ProviderInstance) -> bool:
    settings = loads_json(instance.settings_json, {})
    return bool(settings.get("use_pkce"))


def oauth_integration_configured(
    *,
    provider_app: ProviderApp,
    provider_instance: ProviderInstance,
    needs_tenant: bool = False,
) -> tuple[bool, str | None]:
    """Same rules as frontend oauthIntegrationConfigured (authorize + token + client rules)."""
    settings = loads_json(provider_instance.settings_json, {})
    tenant_ok = True
    if needs_tenant:
        tenant_ok = bool(str(settings.get("tenant_id") or "").strip())
    if not tenant_ok:
        return False, "Directory (tenant) not set"

    authz = str(provider_instance.authorization_endpoint or "").strip()
    tok = str(provider_instance.token_endpoint or "").strip()
    pkce = _use_pkce(provider_instance)

    dcr = bool(getattr(provider_app, "oauth_dynamic_client_registration_enabled", False))
    reg = str(getattr(provider_app, "oauth_registration_endpoint", None) or "").strip()

    if not authz:
        return False, "Authorize URL missing"
    if not tok:
        return False, "Token URL missing"

    if dcr:
        if not reg:
            return False, "Registration URL missing"
        return True, None

    if not str(provider_app.client_id or "").strip():
        return False, "Client ID missing"
    if not bool(provider_app.encrypted_client_secret) and not pkce:
        return False, "Client secret or PKCE required"
    return True, None


def oauth_integration_configured_for_apps(
    provider_app: ProviderApp,
    provider_instance: ProviderInstance | None,
    *,
    needs_tenant: bool = False,
) -> tuple[bool, str | None]:
    if not provider_instance:
        return False, "Provider instance missing"
    return oauth_integration_configured(
        provider_app=provider_app,
        provider_instance=provider_instance,
        needs_tenant=needs_tenant,
    )
